app/src/main/java/software/amazon/freertos/demo/DeviceScanFragment.java

```diff
@@ -67,63 +69,67 @@ public BleDeviceHolder(LayoutInflater inflater, ViewGroup parent) {
             mMenuTextView = (TextView) itemView.findViewById(R.id.menu_option);
         }
 
-        public void bind(BleDevice bleDevice) {
-            mBleDevice = bleDevice;
-            mBleDeviceNameTextView.setText(mBleDevice.getName());
-            mBleDeviceMacTextView.setText(mBleDevice.getMacAddr());
-            mBleDeviceSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
-                AmazonFreeRTOSDevice aDevice = null;
-                boolean autoReconnect = true;
-                @Override
-                public void onCheckedChanged(CompoundButton v, boolean isChecked) {
-                    Log.i(TAG, "Connect switch isChecked: " + (isChecked ? "ON":"OFF"));
-                    if (isChecked) {
-                        if (aDevice == null) {
-                            AWSCredentialsProvider credentialsProvider = AWSMobileClient.getInstance();
-
-                            aDevice = mAmazonFreeRTOSManager.connectToDevice(mBleDevice.getBluetoothDevice(),
-                                    connectionStatusCallback, credentialsProvider, autoReconnect);
-                        }
-                    } else {
-                        if (userDisconnect || !autoReconnect) {
-                            if (aDevice != null) {
-                                mAmazonFreeRTOSManager.disconnectFromDevice(aDevice);
-                                aDevice = null;
-                            }
+        private final CompoundButton.OnCheckedChangeListener deviceConnectListener =
+                new CompoundButton.OnCheckedChangeListener() {
+                    @Override
+                    public void onCheckedChanged(CompoundButton v, boolean isChecked) {
+                        Log.i(TAG, "Connect switch isChecked: " + (isChecked ? "ON" : "OFF"));
+                        if (isChecked) {
+                            connectDevice();
```

Comment:
What if the `connectDevice` attempt fails? If connection attempt fails, doesn't that mean a BLE_DISCONNECT_EVENT won't occur because the connection was never created? In which case, the switch would reflect `ON` despite connection not being present?

Perhaps there should be a conditional `resetUI()` after this, in case the `connectDevice` fails

Reply:
Good point. I verified this. `connect()` is an Async call which returns immediately, but performs a background task to connect to that device. It returns a `DISCONNECTED` callback after a timeout which clears the UI.
So a return value is not necessary here to reset UI.